File: leetcode/202. Happy Number/202._Happy_Number_2.java
public class Solution {
    public boolean isHappy(int n) {
        HashSet<Integer> trace = new HashSet<Integer>();
        int temp = convert(n);
        while( !trace.contains(temp) ) {
            trace.add(temp);
            n = temp;
            temp = convert(n);
            if( temp == 1)
                return true;
        }
        return false;
    }
    public int convert(int n) {
        int ans = 0;
        while( n > 0 ) {
            ans += ((n % 10)*(n % 10));
            n /= 10;
        }
        return ans;
    }
}

File: leetcode/202. Happy Number/202._Happy_Number_2.py
class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        self.trace = []
        return self.check(n)
                
    def check(self, n):
        if n in self.trace:
            #print 'n = ', n, 'trace = ', self.trace
            return False
        else:
            self.trace.append(n)
        sum = 0
        while n > 0:
            digit = n % 10
            n /= 10
            sum += digit ** 2
        if sum == 1:
            return True
        else:
            return self.check(sum)